src/graph/transform/to_simple.cc

```diff
@@ -0,0 +1,100 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file graph/transform/to_simple.cc
+ * \brief Convert multigraphs to simple graphs
+ */
+
+#include <dgl/base_heterograph.h>
+#include <dgl/transform.h>
+#include <dgl/array.h>
+#include <dgl/packed_func_ext.h>
+#include <vector>
+#include <utility>
+#include "../unit_graph.h"
+#include "../../c_api_common.h"
+
+namespace dgl {
+
+using namespace dgl::runtime;
+using namespace dgl::aten;
+
+namespace transform {
+
+std::tuple<HeteroGraphPtr, std::vector<IdArray>, std::vector<IdArray>>
+ToSimpleGraph(const HeteroGraphPtr graph) {
+  const int64_t num_etypes = graph->NumEdgeTypes();
+  const auto metagraph = graph->meta_graph();
+
+  std::vector<IdArray> counts(num_etypes), edge_maps(num_etypes);
+  std::vector<HeteroGraphPtr> rel_graphs(num_etypes);
+
+  for (int64_t etype = 0; etype < num_etypes; ++etype) {
+    const auto vtypes = graph->GetEndpointTypes(etype);
+    const COOMatrix adj = graph->GetCOOMatrix(etype);
+    const COOMatrix sorted_adj = COOSort(adj, true);
+    const IdArray eids_shuffled = sorted_adj.data;
+    const COOMatrix coalesced_adj = COOCoalesce(sorted_adj);
+    const IdArray &count = coalesced_adj.data;
+
+    /*
+     * eids_shuffled actually already contains the mapping from old edge space to the
+     * new one:
+     *
+     * * eids_shuffled[0:count[0]] indicates the original edge IDs that coalesced into new
+     *   edge #0.
+     * * eids_shuffled[count[0]:count[0] + count[1]] indicates those that coalesced into
+     *   new edge #1.
+     * * eids_shuffled[count[0] + count[1]:count[0] + count[1] + count[2]] indicates those
+     *   that coalesced into new edge #2.
+     * * etc.
+     *
+     * Here, we need to translate eids_shuffled to an array "eids_remapped" such that
+     * eids_remapped[i] indicates the new edge ID the old edge #i is mapped to.  The
+     * translation can simply be achieved by (in numpy code):
+     *
+     *     new_eid_for_eids_shuffled = np.range(len(count)).repeat(count)
+     *     eids_remapped = np.zeros_like(new_eid_for_eids_shuffled)
+     *     eids_remapped[eids_shuffled] = new_eid_for_eids_shuffled
+     */
+    const IdArray new_eids = Range(
+        0, coalesced_adj.row->shape[0], coalesced_adj.row->dtype.bits, coalesced_adj.row->ctx);
+    const IdArray eids_remapped = Scatter(Repeat(new_eids, count), eids_shuffled);
```

Comment:
If `COOCoalesce` follows the convention about data array, this part could be avoided.

Reply:
Why can this be avoided?  We still need the mapping from the original graph's edges to those of the new simple graph.

Reply:
I see. Is this part for the reverse mapping from original graph to the result graph (thus having length equal to |E_orig|)? If so, it cannot be avoided.

Reply:
Yes.